### ISSUE TITLE
Token is refreshed once [caused by interceptors' stack movement]

### ISSUE THREAD

**Issue by terapepo** [Flyrell/axios-auth-refresh#12]
So I implemented the library like this:

```
import axios from 'axios';
import { loadProgressBar } from 'axios-progress-bar';
import createAuthRefreshInterceptor from 'axios-auth-refresh';
import 'axios-progress-bar/dist/nprogress.css';
import { apiUrl, headers } from '../services/constants';

const refreshAuthLogic = failedRequest =>
  AXIOS.get('/auth/refresh-token', { headers: headers() }).then(res => {
    sessionStorage.setItem('WinToken', res.data.token);
    failedRequest.response.config.headers['Authorization'] =
      'Bearer ' + res.data.token;
    return Promise.resolve();
  });

const AXIOS = axios.create({
  baseURL: apiUrl,
  headers: {
    'Content-Type': 'application/json',
  },
  timeout: 100000,
});

createAuthRefreshInterceptor(AXIOS, refreshAuthLogic);

AXIOS.interceptors.response.use(
  response => {
    return response;
  },
  error => {
    if (error && error.response && error.response.status === 403) {
      window.location.href = '/';
      return Promise.reject(error);
    }
  },
);

loadProgressBar(undefined, AXIOS);
export default AXIOS;

```
And it works the first time I get a 401 response, but after several tests I confirmed that after this first refresh, it won't refresh the token anymore, the interceptor doesn't trigger again when I receive a 401

**Comment by ventralnet** [comment]
I looked at the source code and it looks like
https://github.com/Flyrell/axios-auth-refresh/blob/553f195fb94c8bb0a1ad8bd2b50f4312ca653a74/src/index.js#L33

on that line the interceptor is removed from the interceptor stack.  This leads me to believe that in your  refreshtoken call, on success you need to add the interceptor back

**Comment by ventralnet** [comment]
Actually nevermind, it looks like it is re-added in the finally block

**Comment by Flyrell** [comment]
The interceptor is applied to the call, which returns some of the statuses specified in `statusCodes` option. That triggers interceptor, which removes itself until the token refresh ends just in case the refreshing fails with any of the statusCode you're intercepting (which would create the inf. loop). In case of success or error original call's handles (`then`/`catch`) should work as expected. When the call ends, interceptor adds itself back. 

@ventralnet What you're trying to say is... if the refresh logic finishes and the original request is called again, you want the interceptor to intercept that for you once again? Wouldn't that be an infinite loop, too? The original call should have fresh and working authentication if the refresh logic's promise resolves, if it doesn't it would be probably be better to reject the promise in refreshing logic as clearly the refreshing failed. Am I not right? Or did I not understand you? Thanks.

**Comment by ventralnet** [comment]
@Flyrell I think the issue arises when you have an interceptor handling errors and forwarding users to an  error page.  So the flow where there would be an issue would be 
1. try to access url
2. 401 access denied token expired
3. axios-auth-refresh interceptor picks up the 401 access denied
4. new auth token acquired
5. url accessed again
**6. Some time passes and the token expires again**
7. try to access another url
8. 401 access denied
9. axios-auth-refresh is now at the bottom of the interceptor stack so Error interceptor handles the 401 and shows error page

I suppose a typical situation where this could occur is if you have a 24 hour expiration token and a user is using your app then leaves for the weekend but doesn't close their browser session and continues work on Monday.

**Comment by an0nEmouse** [comment]
I'm still facing this issue where it only exchanges tokens once and would lead to an error on the next request when the tokens expire again(60 seconds expiry). Could you guide me through how it should be done?

**Comment by Flyrell** [comment]
@an0nEmouse @ventralnet Sorry guys, have been in and out for so long and didn't actually have the time to look at this one. I'll try to do my best in the next few days, but if you're willing to help, I'd be more than happy to see new contributors. @ventralnet You're probably right, but I'm not sure how to add it on top of the error interceptor. Do we need to remove it as well and then re-add it (this sounds like a hack)? or does axios have something add onto the specified index in stack?

**Comment by AngelOD** [comment]
I decided to solve this using a simple boolean that checks if a refresh request is already being handled, and if so it skips the processing and returns the error.

It works, but it's kinda hacky, so I don't think it's worth making a pull request for, but if people disagree, then obviously I can make one.

**Comment by deguich** [comment]
I realise this issue is probably closed than mine #28 
Indeed, a simple "disable refresh process by boolean" could be a solution ...

**Comment by deguich** [comment]
I try with this code (it seems to work) :

```javascript
/** @type {Object} */
const defaults = {

    /** @type {Number[]} */
    statusCodes: [
        401 // Unauthorized
    ]
};

let refreshProcessEnable = true;

/**
 * Creates an authentication refresh interceptor that binds to any error response.
 * If the response code is 401, interceptor tries to call the refreshTokenCall which must return a Promise.
 * While refreshTokenCall is running, all new requests are intercepted and waiting for it to resolve.
 * After Promise is resolved/rejected the authentication refresh interceptor is revoked.
 * @param {AxiosInstance|Function} axios - axios instance
 * @param {Function} refreshTokenCall - refresh token call which must return a Promise
 * @param {Object} options - options for the interceptor @see defaultOptions
 * @return {AxiosInstance}
 */
function createAuthRefreshInterceptor (axios, refreshTokenCall, options = {}) {
    const id = axios.interceptors.response.use(res => res, error => {
        if (!refreshProcessEnable)
            return Promise.reject(error);

        // Reject promise if the error status is not in options.ports or defaults.ports
        const statusCodes = options.hasOwnProperty('statusCodes') && options.statusCodes.length
            ? options.statusCodes
            : defaults.statusCodes;
        if (!error.response || (error.response.status && statusCodes.indexOf(+error.response.status) === -1)) {
            return Promise.reject(error);
        }

        refreshProcessEnable = false;

        // Remove the interceptor to prevent a loop
        // in case token refresh also causes the 401
        //axios.interceptors.response.eject(id);

        const refreshCall = refreshTokenCall(error);

        // Create interceptor that will bind all the others requests
        // until refreshTokenCall is resolved
        const requestQueueInterceptorId = axios.interceptors
            .request
            .use(request => refreshCall.then(() => request));
        // When response code is 401 (Unauthorized), try to refresh the token.
        return refreshCall.then(() => {
            axios.interceptors.request.eject(requestQueueInterceptorId);
            return axios(error.response.config);
        }).catch(error => {
            axios.interceptors.request.eject(requestQueueInterceptorId);
            return Promise.reject(error)
        }).finally(() => {
            //createAuthRefreshInterceptor(axios, refreshTokenCall, options);
            refreshProcessEnable = true;
        });
    });
    return axios;
}
export default createAuthRefreshInterceptor;
```

**Comment by Flyrell** [comment]
This should now be working in beta (but still needs the unit tests). You can try it out:
```
npm install axios-auth-refresh@beta
```

Updated documentation is available in #27. I'd really appreciate the feedback, thank you.

**Comment by SudoPlz** [comment]
@Flyrell looks like the following line https://github.com/Flyrell/axios-auth-refresh/blob/68e8bab35680d7e197fff21be968f1f6f4a13122/src/index.ts#L144

may run if for example the refresh token has expired, thus making the refresh call fail, but no-one would know the specifics of why it failed (a code, or an error message or even custom data passed in the axios config), because the thrown error is too generic.

Am I missing something here?
Thanks

**Comment by Flyrell** [comment]
@SudoPlz The line you highlighted cancels the requests that are waiting for the refresh call to end. The refresh call should throw the original error config. 

I'm wondering if we can throw something here, that looks more like the error object, but with some custom property that would indicate it's been cancelled. I think it might enable users to create only one error handler and handle all the errors similary as they'd have the same structure. Any thoughts so far?

**Comment by SudoPlz** [comment]
Oh got it, I stand corrected, thank you.

**Comment by gerardojbaez** [comment]
@Flyrell I added  `axios-auth-refresh@beta` to my project and configured it with: 

```javascript
createAuthRefreshInterceptor(axios, () => auth.tokenRefresh());
```

Everything is working as expected! Thank you for the library.

**Comment by gerardojbaez** [comment]
Seems like the issue is still happening on the beta branch. The token is refreshed properly on the first try, then fails on subsequent refreshes.

**Comment by Flyrell** [comment]
@gerardojbaez Thanks for the kind words. I will look at the issue, but first, would you please provide me with your axios version? Thanks.

**Comment by gerardojbaez** [comment]
@Flyrell You are welcome! I'm using v0.18.1.

I managed to fix the issue on my case by removing an old interceptor that was handling `401` responses and redirecting the user to the login page. Now, every time a `401` response is returned by the server, `axios-auth-refresh` handles it as expected, all the time; no matter where it is located on the interceptor stack. 

I've added another interceptor that checks for `401` responses on the token refresh endpoint only (the endpoint used on the refresh logic) and then redirects the user to the login page. 

This is how my code looks like:

**Refresh token**
```javascript
createAuthRefreshInterceptor(axios, auth.tokenRefresh)
```

**Redirect to login**
```javascript
axios.interceptors.response.use(function (response) {
    return response
}, function (error) {
    if (error.response.data.error.code === 'token_mismatch') {
        window.location = '/login'
    }

    return Promise.reject(error)
})
```

`token_mismatch` is the error code returned by the token refresh endpoint.
